File: backend/schemas/macro.py
"""Pydantic schemas for macroeconomic data endpoints."""

from datetime import date

from pydantic import BaseModel


class SeriesDataPoint(BaseModel):
    """Single date/value observation."""

    date: date
    value: float


class SeriesResponse(BaseModel):
    """Full series payload returned to the client."""

    series_id: str
    display_name: str
    unit: str
    data: list[SeriesDataPoint]


class YieldCurvePoint(BaseModel):
    """One maturity point on the yield curve."""

    maturity: str
    rate: float


class YieldCurveResponse(BaseModel):
    """Snapshot of the Treasury yield curve for a given date."""

    date: date
    points: list[YieldCurvePoint]


class IndicatorSummary(BaseModel):
    """Summary of a key macroeconomic indicator."""

    name: str
    value: float
    change: float | None = None
    unit: str


# ── Catalog ──


class SeriesCatalogEntry(BaseModel):
    """Describes one available series for the selector UI."""

    series_id: str
    display_name: str
    unit: str
    frequency: str
    category: str
    observation_count: int
    first_date: date | None = None
    last_date: date | None = None


# ── Multi-series ──


class MultiSeriesRequest(BaseModel):
    """Request body for fetching multiple series aligned by date."""

    series_ids: list[str]
    start: date | None = None
    end: date | None = None
    normalize: bool = False  # Z-score normalize for comparison


class SeriesColumn(BaseModel):
    """One column of a multi-series response."""

    series_id: str
    display_name: str
    unit: str
    values: list[float | None]


class MultiSeriesResponse(BaseModel):
    """Multiple series aligned on a common date index."""

    dates: list[date]
    series: list[SeriesColumn]


# ── Correlation ──


class CorrelationRequest(BaseModel):
    """Request body for cross-correlation analysis."""

    series_ids: list[str]
    start: date | None = None
    end: date | None = None
    max_lag: int = 12  # months


class CorrelationPair(BaseModel):
    """Cross-correlation between two series at the optimal lag."""

    series_a: str
    series_b: str
    correlation: float
    optimal_lag: int  # positive = A leads B


class CorrelationMatrix(BaseModel):
    """Pairwise contemporaneous correlations."""

    series_ids: list[str]
    display_names: list[str]
    matrix: list[list[float]]


class CorrelationResponse(BaseModel):
    """Full cross-correlation analysis result."""

    contemporaneous: CorrelationMatrix
    lagged: list[CorrelationPair]


# ── Composite / Recession risk ──


class RecessionSignal(BaseModel):
    """Individual recession risk signal."""

    name: str
    series_id: str
    signal: bool
    value: float
    threshold: str
    description: str


class RecessionRiskResponse(BaseModel):
    """Composite recession risk assessment."""

    score: int  # 0-10
    total_signals: int
    signals: list[RecessionSignal]

File: frontend/src/api/macro.ts
import client from "./client";
import type {
  MacroSeriesResponse,
  YieldCurveResponse,
  SeriesCatalogEntry,
  MultiSeriesRequest,
  MultiSeriesResponse,
  CorrelationRequest,
  CorrelationResponse,
  RecessionRiskResponse,
  IndicatorSummary,
} from "@/types";

export async function getSeries(
  seriesId: string,
  params?: { start_date?: string; end_date?: string }
): Promise<MacroSeriesResponse> {
  const { data } = await client.get<MacroSeriesResponse>(
    `/macro/series/${seriesId}`,
    { params }
  );
  return data;
}

export async function getYieldCurve(
  date?: string
): Promise<YieldCurveResponse> {
  const { data } = await client.get<YieldCurveResponse>("/macro/yield-curve", {
    params: date ? { date } : undefined,
  });
  return data;
}

export async function getIndicators(): Promise<IndicatorSummary[]> {
  const { data } = await client.get<IndicatorSummary[]>("/macro/indicators");
  return data;
}

export async function getCatalog(): Promise<SeriesCatalogEntry[]> {
  const { data } = await client.get<SeriesCatalogEntry[]>("/macro/catalog");
  return data;
}

export async function getMultiSeries(
  body: MultiSeriesRequest
): Promise<MultiSeriesResponse> {
  const { data } = await client.post<MultiSeriesResponse>(
    "/macro/multi-series",
    body
  );
  return data;
}

export async function getCorrelation(
  body: CorrelationRequest
): Promise<CorrelationResponse> {
  const { data } = await client.post<CorrelationResponse>(
    "/macro/correlation",
    body
  );
  return data;
}

export async function getRecessionRisk(): Promise<RecessionRiskResponse> {
  const { data } = await client.get<RecessionRiskResponse>(
    "/macro/recession-risk"
  );
  return data;
}
